algorithms/tree_algorithm.py

```diff
@@ -0,0 +1,187 @@
+        if count_flag:
+            return 0
+        return False
+
+    # If there are two roots, the amount of isomorphisms is the addition of isomorphisms from the two mappings
+    # of roots to each other. Because the maximum amount of roots is 2, all roots of T1 are mapped to one root of T2.
+    # Root levels are overwritten in the second cycle.
+    total_isomorphisms = 0
+    # for i_root in range(len(root_T1)):
+    for i_root in [0]:
+        root1 = root_T1[i_root]
+        root2 = root_T2[0]
+
+        # Assign level numbers to all nodes
+        __assign_level(root1, 0, [])
+        __assign_level(root2, 0, [])
+
+        # Assign all vertices to level number lists
+        L1 = {}
+        for v in T1:
+            L1.setdefault(v.level, []).append(v)
+        L2 = {}
+        for v in T2:
+            L2.setdefault(v.level, []).append(v)
+
+        # The number of levels of each tree should be equal, otherwise not isomorphic
+        h1, h2 = max(L1.keys()), max(L2.keys())
+        if max(L1.keys()) != max(L2.keys()):
+            continue
+        h = h1
+
+        # From the bottom level up, assign names to the vertices
+        # If the vertex is a leaf, the name 'lr' is assigned
+        # The name of all other vertices is a sorted collection of the names of its children
+        # At the same time, remember how many automorphisms the subtree of that vertex has
+        are_isomorphic = True    # For each level it is checked if the trees still can be isomoprhic, if not, the loop can be stopped
+        for i in range(h, -1, -1):
+            # Name the vertices of tree T1
+            H1 = []
+            for v in L1[i]:
+                if v.degree_fixed == 1:
+                    v.name = 'lr'
+                    v.auto = 1
+                else:
+                    # For each vertex, get the set of string assigned to its children
+                    children_names = []
+                    children_count = {}
+                    children_auto = {}
+                    for n in v.neighbours:
+                        if n.level > v.level:
+                            for _ in range(n.n_twins):
+                                children_names.append(n.name)
+                            children_count[n.name] = children_count.setdefault(n.name, 0) + 1
+                            children_auto[n.name] = children_auto.setdefault(n.name, 1) * n.auto
+                    v_auto = 1
+                    for child_name in set(children_names):
+                        v_auto *= factorial(children_count[child_name]) * children_auto[child_name]
+                    v.auto = v_auto
+
+                    children_names.sort()
+                    v_name = 'l'
+                    for child_name in children_names:
+                        v_name += child_name
+                    v_name += 'r'
+                    v.name = v_name
+                H1.append(v.name)
+
+            # Name the vertices of tree T2
```

Comment:
Zou je dit misschien als een functie kunnen maken waar dan L1 of L2 ingaat en H1 of H2 uitgaat? Dat scheelt een bult dubbele code volgens mij.

Reply:
maakt het voor mij ook leesbaarder

Reply:
Goed plan, ga ik doen!